clearblade.go

```diff
@@ -113,57 +120,35 @@ func fetchAdapterConfig() (*AdapterConfig, error) {
-		query.EqualTo("adapter_name", Args.DeviceName)
-	}
+func PublishStatus(topic string, json []byte) (mqtt.Token, error) {
+	log.Printf("[INFO] PublishStatus - Setting status on topic: %s\n", topic)
 
-	log.Println("[DEBUG] setConnectionStatus - Executing query against table " + Args.AdapterConfigCollection)
-	results, err := deviceClient.GetDataByName(Args.AdapterConfigCollection, query)
+	log.Printf("[DEBUG] PublishStatus - Publishing to topic %s\n", topic)
+	token, err := PublishGetToken(topic, json)
 	if err != nil {
-		log.Printf("[ERROR] setConnectionStatus - Error retrieving adapter configuration: %s\n", err.Error())
-		log.Println("[ERROR] setConnectionStatus - Retrying in 30 seconds...")
-		time.Sleep(time.Second * 30)
-		SetConnectionStatus(status)
+		log.Printf("[ERROR] Failed to publish MQTT message to topic %s: %s\n", topic, err.Error())
 	}
-	data := results["DATA"].([]interface{})
-	changes := make(map[string]interface{})
-	if len(data) > 0 {
-		log.Println("[INFO] setConnectionStatus - Adapter config retrieved")
 
-		changes["connection_status"] = status
-
-		response, err := deviceClient.UpdateDataByName(Args.SystemKey, Args.AdapterConfigCollection, query, changes)
-		log.Printf("[INFO] setConnectionStatus - Updated %.0f row(s) in collection: %s", response.Count, Args.AdapterConfigCollection)
-		if err != nil {
-			log.Printf("[ERROR] setConnectionStatus - Error updating adapter configuration: %s\n", err.Error())
-			log.Println("[ERROR] setConnectionStatus - Retrying in 30 seconds...")
-			time.Sleep(time.Second * 30)
-			SetConnectionStatus(status)
-		}
-	}
-
-	log.Printf("[DEBUG] setConnectionStatus - Successfully received, parsed, and updated adapter config: %+v\n", changes)
-	return nil
+	log.Printf("[DEBUG] PublishStatus - Successfully sent status message to topic: %s\n", topic)
+	return token, err
 }
 
 func onConnectLost(client mqtt.Client, connerr error) {
-	log.Printf("[ERROR] onConnectLost - Connection to MQTT broker was lost: %s\n", connerr.Error())
-	if Args.ServiceAccount == "" {
-		log.Fatalln("[FATAL] onConnectLost - MQTT Connection was lost and no Device Service Account is being used. Stopping Adapter to force device reauth (this can be avoided by using Device Service Accounts)")
+	if Args.FatalOnDisconnect == "true" {
+		log.Fatalf("[FATAL] onConnectLost - Connection to MQTT broker was lost: %s\n", connerr.Error())
+		if Args.ServiceAccount == "" {
+			log.Fatalln("[FATAL] onConnectLost - MQTT Connection was lost and no Device Service Account is being used. Stopping Adapter to force device reauth (this can be avoided by using Device Service Accounts)")
+		}
+	} else {
+		log.Printf("[ERROR] onConnectLost - Connection to MQTT broker was lost: %s\n", connerr.Error())
+		if Args.ServiceAccount == "" {
+			log.Fatalln("[FATAL] onConnectLost - MQTT Connection was lost and no Device Service Account is being used. Stopping Adapter to force device reauth (this can be avoided by using Device Service Accounts)")
+		}
 	}
```

Comment:
What is the purpose of having the FatalOnDisconnect flag if we are going to stop the adapter in either case?

Reply:
Disregard the comment above. I see that when FatalOnDisconnect is false, we only end the adapter when a service account is not being used. However, if we do not kill the adapter, how will the MQTT connection be re-established? I don't see anywhere that we are attempting to reconnect.

Reply:
I was just maintaining the old logic here. That is how it behaved before I made any changes. I'm open to making it act however we want.

Reply:
Great call out Jim - I've got on my to-do list to add MQTT reconnect logic here, because as of now nothing happens and the adapter gets in a stale state. On the reconnect side though we need to track the topics we are subscribed to so that we can automatically resubscribe, or the other option is to provide a callback to our adapter when MQTT is disconnected so that it can reconnect and resubscribe as needed.